numpy/_core/src/multiarray/scalartypes.c.src

```diff
@@ -3922,7 +3922,9 @@ static npy_hash_t
     dtype = PyArray_DescrFromScalar(obj);
     meta = get_datetime_metadata_from_dtype(dtype);
```

Comment:
While we're at it, let's also fix two other bugs in this function. Both of these calls can fail, so we need to check for errors.

```suggestion
    dtype = PyArray_DescrFromScalar(obj);
    if (dtype == NULL) {
        return -1;
    }
    meta = get_datetime_metadata_from_dtype(dtype);
    if (meta == NULL) {
        Py_DECREF(dtype);
        return -1;
    }
```

Reply:
Oh yes, good spot! Thanks.

Reply:
Actually, please just get the metadata directly from the scalar. Then it also cannot fail.

Reply:
good point! I edited my suggestion. @kostayScr if you're interested in getting this across the finish line, you'll need to apply my suggestions and also delete the now-unnecessary declaration of `dtype` outside the context of the PR, since I can't suggest that from the PR interface.